lib/update/clusterconfig/phases/k8s.go

```diff
@@ -0,0 +1,167 @@
+		}
+	})
+}
+
+func createServiceWithClusterIP(ctx context.Context, service v1.Service, alloc *ipallocator.Range, services corev1.ServiceInterface, logger log.FieldLogger) error {
+	ip, err := alloc.AllocateNext()
+	if err != nil {
+		return trace.Wrap(err, "failed to allocate service IP")
+	}
+	return utils.RetryWithInterval(ctx, newOperationBackoff(), func() error {
+		service.Spec.ClusterIP = ip.String()
+		logger.WithField("cluster-ip", service.Spec.ClusterIP).Info("Recreate service with cluster IP.")
+		service.ResourceVersion = "0"
+		_, err = services.Create(&service)
+		if err == nil || errors.IsAlreadyExists(err) {
+			return nil
+		}
+		switch {
+		case utils.IsTransientClusterError(err), isIPRangeMistmatchError(err):
+			// Fall-through
+		case isIPAlreadyAllocatedError(err):
+			alloc.Release(ip)
+			ip, err = alloc.AllocateNext()
+			if err != nil {
+				return &backoff.PermanentError{Err: err}
+			}
+			// Fall-through
+		default:
+			return &backoff.PermanentError{Err: err}
+		}
+		// Retry on transient errors
+		return rigging.ConvertError(err)
+	})
+}
+
+// isIPRangeMistmatchError detects whether the given error indicates that the suggested cluster IP
+// is from an unexpected service IP range. This can happen as long as the apiserver's repair
+// step did not commit the new service IP range configuration to the store (eg etcd)
+func isIPRangeMistmatchError(err error) bool {
+	switch err := err.(type) {
+	case *errors.StatusError:
+		return err.ErrStatus.Status == "Failure" && statusHasCause(err.ErrStatus,
+			"spec.clusterIP", "provided range does not match the current range")
+	}
+	return false
+}
+
+// isIPAlreadyAllocatedError detects whether the given error indicates that the specified
+// cluster IP is already allocated.
+// This can happen since we are not syncing the IP allocation with the apiserver
+func isIPAlreadyAllocatedError(err error) bool {
+	switch err := err.(type) {
+	case *errors.StatusError:
+		return err.ErrStatus.Status == "Failure" && statusHasCause(err.ErrStatus,
+			"spec.clusterIP", "provided IP is already allocated")
+	}
+	return false
+}
+
+func statusHasCause(status metav1.Status, field, messagePattern string) bool {
+	if status.Details == nil {
+		return false
+	}
+	for _, cause := range status.Details.Causes {
+		if cause.Field == field && strings.Contains(cause.Message, messagePattern) {
+			return true
+		}
+	}
+	return false
+}
+
+func newOperationBackoff() backoff.BackOff {
+	return utils.NewExponentialBackOff(5 * time.Minute)
+}
+
+// shouldManageService decides whether the configuration operation should
+// manage the given service.
+// The operation will manage a service, if it's of type clusterIP and is not a headless service
+// (i.e. has a valid IP address assigned from the service subnet). Additionally, DNS services
+// as well as the API server service are managed elsewhere and hence excluded
+func shouldManageService(service v1.Service) bool {
```

Comment:
It sounds to me that this function should rather be called should**Not**ManageService? Cause it returns true for all services that its description says should not be managed.